### PR TITLE
FIX Improve non-inline validation

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-elemental/issues/1155

Need to merge https://github.com/silverstripe/silverstripe-framework/pull/11213 first to get the has_one RequiredFields validation working correctly and https://github.com/silverstripe/silverstripe-frameworktest/pull/172 to get behat to pass.

There is a mutli-pr CI run linked on the issue

Note that this PR does not touch inline-saving

**Scenarios:**
1. DataObject validate() addFieldError()
2. DataObject validate() arrError()
3. UrlField validation
4. RequiredFields on default Title field
5. has_one relation to File and RequiredFields
6. has_one relation to SiteTree and RequiredFields

**Non-inline editable block:**
1. Already worked
2. Already worked
3. Already worked
4. Fixed in this PR - by default it's a composite field and validation works but doesn't show message on front-end
5. Already worked
6. Fixed in framework PR

**Inline-editable block page save**
1. Already worked - UX is bad not fixed in this PR - [issue](https://github.com/silverstripe/silverstripe-elemental/issues/1179)
2. Already worked - UX is bad not fixed in this PR - [issue](https://github.com/silverstripe/silverstripe-elemental/issues/1179)
3. Fixed in this PR
4. Fixed in this PR
5. Already worked
6. Fixed in framework PR

I used the following blocks to tests these scenarios

```php
<?php

use DNADesign\Elemental\Models\BaseElement;
use SilverStripe\Forms\CompositeValidator;
use SilverStripe\Forms\RequiredFields;
use SilverStripe\Forms\UrlField;
use SilverStripe\Assets\Image;
use SilverStripe\CMS\Model\SiteTree;

class MyBlock extends BaseElement
{
    private static $db = [
        'MyField' => 'Varchar',
        'MyHTML' => 'HTMLText',
        'MyUrl' => 'Varchar',
    ];

    private static $has_one = [
        'MyImage' => Image::class,
        'MyPage' => SiteTree::class,
    ];

    private static $table_name = 'MyBlock';

    private static $singular_name = 'My Block';

    private static $plural_name = 'My Blocks';

    private static $description = 'This is my block';

    private static $icon = 'font-icon-block-content';

    public function getType()
    {
        return 'My Block';
    }

    public function validate()
    {
        $result = parent::validate();
        if ($this->Title == 'x') {
            $result->addFieldError('Title', 'Title cannot be x', 'validation');
        }
        if ($this->MyField == 'x') {
            $result->addFieldError('MyField', 'MyField cannot be x', 'validation');
        }
        $c = trim(strip_tags($this->MyHTML ?? ''));
        if ($c == 'x') {
            $result->addFieldError('MyHTML', 'MyHTML cannot be x', 'validation');
        }
        if ($this->MyField == 'z' && $c == 'z') {
            $result->addError('This is a general error message');
        }
        return $result;
    }

    public function getCMSCompositeValidator(): CompositeValidator
    {
        return CompositeValidator::create([RequiredFields::create(['Title', 'MyImage', 'MyPageID'])]);
    }

    public function getCMSFields()
    {
        $fields = parent::getCMSFields();
        $fields->dataFieldByName('MyHTML')->setRows(5);
        $fields->removeByName('MyUrl');
        // URLfield contains its own validate() methods
        $fields->addFieldToTab('Root.Main', UrlField::create('MyUrl', 'My URL'));
        return $fields;
    }
}
```

```php
<?php

class MyNonInlineBlock extends MyBlock
{
    private static $inline_editable = false;

    private static $table_name = 'MyNonInlineBlock';

    private static $singular_name = 'My non-inline Block';

    private static $plural_name = 'My non-inline Blocks';

    private static $description = 'This is my non-inline block';

    private static $icon = 'font-icon-block-content';

    public function getType()
    {
        return 'My non-inline Block';
    }
}
```

### CODE REVIEW

File: src/Forms/TextCheckboxGroupField.php
Comment (emteknetnz) [inline src/Forms/TextCheckboxGroupField.php]: TextCheckboxGroupField is the special elemental field that is a composite field of the default `Title` and `ShowTitle` fields

File: src/Forms/ElementalAreaField.php
Comment (emteknetnz) [inline src/Forms/ElementalAreaField.php:224]: Previously FormField validation wasn't happening on fields on elements e.g. `$db = [ 'MyUrl' => UrlField::class ]` when saving the page (as opposed to editing the element directly)

File: tests/Behat/features/inline-block-validation.feature
Comment (GuySartorelli) [inline tests/Behat/features/inline-block-validation.feature]: If we need comments like this separating out scenarios, it's probably better to just create a new feature file instead.

File: src/Models/BaseElement.php
Comment (GuySartorelli) [inline src/Models/BaseElement.php]: Don't leave commented-out code

File: src/Forms/TextCheckboxGroupField.php
Comment (GuySartorelli) [reply src/Forms/TextCheckboxGroupField.php]: These methods feel like they belong on `CompositeField` directly. Please move them there, and include some basic unit tests.

File: tests/Behat/features/non-inline-block-validation.feature
Comment (GuySartorelli) [inline tests/Behat/features/non-inline-block-validation.feature]:
```suggestion
    And content blocks are not in-line editable
```

File: tests/Behat/features/non-inline-block-validation.feature
Comment (GuySartorelli) [inline tests/Behat/features/non-inline-block-validation.feature]: Can these be split up into separate scenarios? If so, please do so since they're clearly each testing a distinct piece of validation as indicated by the comments.

Comment (GuySartorelli) [reply tests/Behat/features/non-inline-block-validation.feature]: Please do the same thing in the other feature file as well, if possible.

File: src/Forms/TextCheckboxGroupField.php
Comment (GuySartorelli) [inline src/Forms/TextCheckboxGroupField.php]: If these methods are inherited from a superclass or interface, we don't need to (and shouldn't, in case it gets stale) duplicate the phpdoc typehints here.

If these are **not** inherited from a superclass or interface, please make these strict typehints instead.

This applies to all such methods in this PR.

File: src/Models/BaseElement.php
Comment (GuySartorelli) [inline src/Models/BaseElement.php]:
```suggestion
```
That's not needed

File: src/Forms/ElementalAreaField.php
Comment (GuySartorelli) [inline src/Forms/ElementalAreaField.php]:
```suggestion
```
Not needed

File: src/Forms/ElementalAreaField.php
Comment (GuySartorelli) [inline src/Forms/ElementalAreaField.php]:
```suggestion
            $formValidatorResult = $elementForm->getValidator()->validate();
```
We should skip the logic in `validationResult()` which might skip validation - at this point we know for sure we want to validate the form, so we should just do it directly.

File: src/Forms/ElementalAreaField.php
Comment (GuySartorelli) [inline src/Forms/ElementalAreaField.php:246]: What do the changes to the `saveInto()` block resolve? I tried changing this method back to what it was before this PR, and it seems to work exactly the same way as before.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
If I save the page without expanding the block, none of the validation seems to happen. e.g. with your test blocks from the description, I add the block and immediately save the page - it should trigger validation errors for the missing required fields but it doesn't.

File: src/Forms/ElementalAreaField.php
Comment (emteknetnz) [reply src/Forms/ElementalAreaField.php]: Updated

File: src/Forms/ElementalAreaField.php
Comment (emteknetnz) [reply src/Forms/ElementalAreaField.php:246]: There's logic I moved to `private function getElementData()` which is called by both the old `saveInto()` method and the new `validate()` method

File: src/Forms/ElementalAreaField.php
Comment (emteknetnz) [reply src/Forms/ElementalAreaField.php]: Removed

File: src/Forms/TextCheckboxGroupField.php
Comment (emteknetnz) [reply src/Forms/TextCheckboxGroupField.php]: I've [raised a new card](https://github.com/silverstripe/silverstripe-framework/issues/11215) for the refactor to do this. I feel like there's already a lot going on in this card including an existing framework PR that does different things, and there's [another issue](https://github.com/silverstripe/silverstripe-elemental/issues/1170) that's dependent on the PRs in this issue

Comment (GuySartorelli) [reply src/Forms/TextCheckboxGroupField.php]: If we keep it in this PR, that means you need to add unit tests here and then to refactor we're removing both the src code and the tests, which shouldn't have been added to this repo in the first place.
Moving them now doesn't introduce any new complexity, it just puts the new code in the right place the first time around.

File: src/Forms/TextCheckboxGroupField.php
Comment (emteknetnz) [reply src/Forms/TextCheckboxGroupField.php]: Inherited, removed

File: src/Models/BaseElement.php
Comment (emteknetnz) [reply src/Models/BaseElement.php]: Removed

File: src/Models/BaseElement.php
Comment (emteknetnz) [reply src/Models/BaseElement.php]: Removed

File: tests/Behat/features/non-inline-block-validation.feature
Comment (emteknetnz) [reply tests/Behat/features/non-inline-block-validation.feature]: Updated

File: tests/Behat/features/inline-block-validation.feature
Comment (emteknetnz) [reply tests/Behat/features/inline-block-validation.feature]: Separated

File: tests/Behat/features/non-inline-block-validation.feature
Comment (emteknetnz) [reply tests/Behat/features/non-inline-block-validation.feature]: Putting them in separate scenarios greatly increases how long it takes for behat to run through the feature file because it needs to rebuild the test database and redo the background steps. Elemental behat speed is already painfully slow and I don't want it taking even longer if possible.

Comment (emteknetnz) [reply tests/Behat/features/non-inline-block-validation.feature]: Had an offline discussion about this, decided it's not worth doing irrespective of performance difference

File: src/Forms/TextCheckboxGroupField.php
Comment (emteknetnz) [reply src/Forms/TextCheckboxGroupField.php]: Like I said, there's already an [existing framework PR ](https://github.com/silverstripe/silverstripe-framework/pull/11213) that deals with RequiredFields that's required for this PR to work.  I don't want to add CompositeFields functionality to that PR because that has nothing to do with RequireFields, so I want to do a seperate PR.  I don't want to block this PR on another framework PR.

It's just easier to refactor this later I don't see the problem

Comment (GuySartorelli) [reply src/Forms/TextCheckboxGroupField.php]: Required fields PR doesn't seem necessary, see comment on that PR.

If that PR does end up being necessary (and please don't comment about that here, comment about that in ***that*** PR), we can keep this here.
If that PR ends up not being necessary, then I think this code should be moved.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
The problem described in https://github.com/silverstripe/silverstripe-elemental/pull/1178#pullrequestreview-2032837575 has not been resolved

File: src/Forms/ElementalAreaField.php
Comment (michalkleiner) [inline src/Forms/ElementalAreaField.php]:
```suggestion
        // that a non-inline edit form will use - see GridFieldDetailForm::handleItem()
```

File: src/Forms/ElementalAreaField.php
Comment (michalkleiner) [inline src/Forms/ElementalAreaField.php]:
```suggestion
        // The validator needs to be explicitly copied from the element to the form
```

File: src/Forms/TextCheckboxGroupField.php
Comment (michalkleiner) [inline src/Forms/TextCheckboxGroupField.php:46]: Should this concat the messages from multiple fields? At the moment it would skip the second message if there were two of them.

File: src/Forms/TextCheckboxGroupField.php
Comment (GuySartorelli) [reply src/Forms/TextCheckboxGroupField.php]: Actually nevermind - after refinment today I'm okay with this being done separately.

**Comment by emteknetnz** [comment]
> If I save the page without expanding the block, none of the validation seems to happen. e.g. with your test blocks from the description, I add the block and immediately save the page - it should trigger validation errors for the missing required fields but it doesn't.

That would be hard to solve.

In the case of inline blocks, currently if they haven't been opened and the page is saved, then then because the child FormBuilder form hasn't been rendered then the element data will not be sent as part of the page form submission.  However if the inline-element is opened then page save will submit element data with the rest of the page, and closing the element then saving the page will also send the element data.  So to solve this we'd need to render the FormBuilder forms for ALL unopened elements on page save

In the case of non-inline blocks, I really don't know what can be done since their forms are never rendered on the page and and thus their data is never sent.

The fact this doesn't work doesn't actually matter given the AC is `Saving/publishing a page will trigger validation on its child block if they are in a a dirty state.` - the elements in question aren't dirty, they're just created in an invalid state.

We could split this off as a seperate card, though I honestly don't know how or even if we should attempt a fix here

File: src/Forms/TextCheckboxGroupField.php
Comment (emteknetnz) [reply src/Forms/TextCheckboxGroupField.php]: [Other card](https://github.com/silverstripe/silverstripe-framework/issues/11215) has now been refined so we'll do on that card

**Comment by GuySartorelli** [comment]
> We could split this off as a seperate card, though I honestly don't know how or even if we should attempt a fix here

I think that's worth looking into. There's a tradeoff to be made between:
1. always having the form rendered (and just hidden until it's expanded) - which means we never have to do workarounds like explicitly render it when hitting the save button on the block
1. potential performance loss from eagerly getting the required data to render the form for all blocks on a given page

I'll open a separate card about this for discussion.

Edit: https://github.com/silverstripe/silverstripe-elemental/issues/1183

File: src/Forms/TextCheckboxGroupField.php
Comment (emteknetnz) [reply src/Forms/TextCheckboxGroupField.php:46]: I think it OK to skip. Templates are setup to only show a single validation message per field, so there would need to be further changes made to support multiple validation messages. In this case of this particular field, this field is a combination of the Title and ShowTitle fields, and since ShowTitle is a boolean checkbox field its impractical to have validation on it.

This code will make its way over to Composite field as part of [this card](https://github.com/silverstripe/silverstripe-framework/issues/11215) - it may make sense to look at multi validation messages there

File: src/Forms/ElementalAreaField.php
Comment (emteknetnz) [reply src/Forms/ElementalAreaField.php]: Updated

File: src/Forms/ElementalAreaField.php
Comment (emteknetnz) [reply src/Forms/ElementalAreaField.php]: Updated

File: src/Forms/TextCheckboxGroupField.php
Comment (GuySartorelli) [reply src/Forms/TextCheckboxGroupField.php:46]: I agree with Michal here, sorry I didn't pick this up in my review, we could have saved some time.

> Templates are setup to only show a single validation message per field

While that is broadly true, this is an edge case where this isn't really *a field* in itself - it's really a holder for multiple fields. If each of the child fields has a validation error message, we should see those.

It would be bad UX to display only one validation message at a time, so users have to fix errors one field at a time, save, and see what fails next.

> since ShowTitle is a boolean checkbox field its impractical to have validation on it.

We can't really decide for developers what they should or shouldn't put validation on, unless we actually lock them out of doing so with exception messages when they try.

## Options
1. Have a new `getMessages()` method, and use that in the template for this form field, loop through the child messages and display all of them
2. Just update this method to concatenate the error messages together - at least the messages will display, even if it's not perfect
3. Explicitly defer this by adding ACs to the follow-up card you made. That's probably best, because there are additional things to consider such as the message type and cast, which may be different for the different messages.

**Review by GuySartorelli — COMMENTED** [review]
Just the one thing left and then we're good, I think.

File: src/Forms/TextCheckboxGroupField.php
Comment (emteknetnz) [reply src/Forms/TextCheckboxGroupField.php:46]: Go with 3, I've updated AC's on the follow up card

**Comment by GuySartorelli** [comment]
Reran CI after merging frameworktest. Confirmed they are running with the correct commit:
`silverstripe/frameworktest                     1.x-dev d12463a`
https://github.com/silverstripe/silverstripe-frameworktest/commit/d12463ae60997ed7f9235b8ea222daba4c4e6413

**Comment by GuySartorelli** [comment]
behat failing

**Comment by emteknetnz** [comment]
I've added an extra code change that's from a [different PR](https://github.com/silverstripe/silverstripe-elemental/pull/1180/files#diff-942115e4b8f6030e4d72ebc2b3a0772ec65e5dfcd08fbd0e677c70d1231daf28R142 ) that's blocked on this one which is required to get the inline-block-validation.feature behat tests to pass

File: src/Controllers/ElementalAreaController.php
Comment (emteknetnz) [inline src/Controllers/ElementalAreaController.php:142]: I've added this code change in from a [different PR](https://github.com/silverstripe/silverstripe-elemental/pull/1180/files#diff-942115e4b8f6030e4d72ebc2b3a0772ec65e5dfcd08fbd0e677c70d1231daf28R142) that's blocked on this one which is required to get the inline-block-validation.feature behat tests to pass

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
I don't know how I didn't notice this before, but for inline-editable blocks when you save the page, only one required field validation message appears at a time.

So if I don't enter any data I only get told about the title, then I add a title and I'm told about the image, then I add an image and I'm told about the page.

The HTTP response does have all three validation errors at the same time - at this stage I don't know if getting them to display is part of this card or not so if there's another card to handle this please just link to it and I'll be okay with that.

**Comment by GuySartorelli** [comment]
Rerunning failed behat tests now that the framework PR is merged, but regardless of whether that goes green I need at least a response to my above comment.

**Comment by GuySartorelli** [comment]
Still need to respond to https://github.com/silverstripe/silverstripe-elemental/pull/1178#pullrequestreview-2041868677

**Comment by emteknetnz** [comment]
Split of as [separate card](https://github.com/silverstripe/silverstripe-elemental/issues/1187)  that can be refined

**Review by GuySartorelli — APPROVED** [review]
LGTM with the understanding that there's still a few issues left to be done before this functionality is considered 100% working 😅

**Pull request merged**
